File: v2/lib/domain/debt.ts
// Hutang & cicilan — port dari server.js (buildDebtSummary + greedy pay/settle).
// Planner murni: balik daftar slot yang harus di-flip jadi paid + carry baru
// + nominal payment yang dicatat. Repo yang eksekusi tulisannya.
//
// Tiga sumber tagihan: main biasa (per pemain per game), iuran patungan
// turnamen (flat, rata ke semua peserta), dan kok per partai turnamen (cuma
// ditagih ke 4 pemain partai itu — beda pasangan main beda jumlah partai,
// jadi tidak adil kalau dirata ke semua peserta lewat fee). Planner
// memperlakukan ketiganya sebagai "slot belum bayar" yang sama, jadi
// cicilan/lunasin-semua otomatis nutup semuanya, dari yang paling pas.

import { gameCost } from "./game";
import {
  matchKokPerPerson,
  matchParticipants,
  matchPlayedDate,
  matchTitle,
  tournamentMatches,
} from "./tournament";
import type {
  CarryMap,
  DebtEntry,
  DebtKind,
  EnrichedGame,
  EnrichedTournament,
  StoredGame,
  StoredTournament,
} from "./types";

/**
 * Partai ke berapa tiap game di hari itu — urut dari yang paling dulu dicatat
 * (createdAt paling lama = Partai 1), biar "Main" di rekap tidak ambigu
 * kalau orang yang sama main lebih dari sekali di hari yang sama.
 */
function gameMatchNumbers(games: EnrichedGame[]): Map<string, number> {
  const byDate = new Map<string, EnrichedGame[]>();
  for (const g of games) {
    const list = byDate.get(g.date);
    if (list) list.push(g);
    else byDate.set(g.date, [g]);
  }
  const out = new Map<string, number>();
  for (const list of byDate.values()) {
    const sorted = [...list].sort((a, b) => a.createdAt.localeCompare(b.createdAt));
    sorted.forEach((g, i) => out.set(g.id, i + 1));
  }
  return out;
}

/** Ringkasan hutang per orang: sisa = max(0, owedGross − carry). */
export function buildDebtSummary(
  games: EnrichedGame[],
  carryMap: CarryMap,
  tournaments: EnrichedTournament[] = [],
): DebtEntry[] {
  const byName: Record<string, Omit<DebtEntry, "carry" | "total">> = {};
  const bucket = (name: string) => {
    if (!byName[name]) byName[name] = { name, owedGross: 0, items: [] };
    return byName[name];
  };
  const matchNumbers = gameMatchNumbers(games);

  for (const g of games) {
    for (const p of g.players) {
      if (p.paid || !p.name) continue;
      const e = bucket(p.name);
      e.owedGross += g.cost.perPerson;
      e.items.push({
        gameId: g.id,
        date: g.date,
        name: p.name,
        amount: g.cost.perPerson,
        kokCount: g.cost.kokCount,
        kind: "game",
        createdAt: g.createdAt,
        matchNumber: matchNumbers.get(g.id),
      });
    }
  }

  for (const t of tournaments) {
    if (t.fee <= 0) continue;
    for (const f of t.fees) {
      if (f.paid || !f.name) continue;
      const e = bucket(f.name);
      e.owedGross += t.fee;
      e.items.push({
        gameId: t.id,
        date: t.date,
        name: f.name,
        amount: t.fee,
        // Iuran flat ini bukan basis kok — kok per partai punya item sendiri di bawah.
        kokCount: 0,
        kind: "turnamen",
        label: t.name,
        createdAt: t.createdAt,
      });
    }
  }

  for (const t of tournaments) {
    for (const m of t.matches) {
      if (m.kokTotal <= 0) continue;
      const perPerson = matchKokPerPerson(m);
      if (perPerson <= 0) continue;
      const names = matchParticipants(m);
      for (let i = 0; i < names.length; i++) {
        const name = names[i];
        if (!name || m.kokPaid[i]) continue;
        const e = bucket(name);
        e.owedGross += perPerson;
        e.items.push({
          gameId: `${t.id}::${m.id}`,
          date: matchPlayedDate(t, m),
          name,
          amount: perPerson,
          kokCount: m.koks.length,
          kind: "turnamen_kok",
          label: `${t.name} · ${matchTitle(t, m)}`,
          createdAt: t.createdAt,
        });
      }
    }
  }

  return Object.values(byName)
    .map((e) => {
      const carry = Math.max(0, Number(carryMap?.[e.name]) || 0);
      return { ...e, carry, total: Math.max(0, e.owedGross - carry) };
    })
    .sort((a, b) => b.total - a.total || a.name.localeCompare(b.name, "id"));
}

export interface TouchedSlot {
  kind: DebtKind;
  /** id game, atau id turnamen kalau kind = "turnamen"/"turnamen_kok". */
  id: string;
  /**
   * index pemain di game (0-3), index peserta di daftar iuran turnamen, atau
   * slot pemain di partai (0-3, urutan [sisiA.a, sisiA.b, sisiB.a, sisiB.b])
   * kalau kind = "turnamen_kok".
   */
  index: number;
  /** Cuma kepake buat kind "turnamen_kok" — id partai yang kok-nya ditagih. */
  matchId?: string;
}

export interface SettlePlan {
  touched: TouchedSlot[];
  /** null = hapus entri carry; number = set carry ke nilai ini. */
  carryAfter: number | null;
  /** nominal yang dicatat di ledger payments. */
  paymentAmount: number;
  /** true kalau setelah bayar ini semua tagihannya habis → dicatat sebagai "lunas". */
  clearsDebt: boolean;
}

interface UnpaidRef extends TouchedSlot {
  date: string;
  createdAt: string;
  amount: number;
}

function unpaidRefs(
  games: StoredGame[],
  name: string,
  tournaments: StoredTournament[] = [],
): UnpaidRef[] {
  const refs: UnpaidRef[] = [];
  for (const g of games) {
    const perPerson = gameCost(g).perPerson;
    for (let i = 0; i < g.players.length; i++) {
      if (g.players[i].name === name && !g.players[i].paid) {
        refs.push({
          kind: "game",
          id: g.id,
          index: i,
          date: g.date,
          createdAt: g.createdAt,
          amount: perPerson,
        });
      }
    }
  }
  for (const t of tournaments) {
    if (t.fee > 0) {
      for (let i = 0; i < t.fees.length; i++) {
        if (t.fees[i].name === name && !t.fees[i].paid) {
          refs.push({
            kind: "turnamen",
            id: t.id,
            index: i,
            date: t.date,
            createdAt: t.createdAt,
            amount: t.fee,
          });
        }
      }
    }
    for (const m of tournamentMatches(t)) {
      if (m.kokTotal <= 0) continue;
      const perPerson = matchKokPerPerson(m);
      if (perPerson <= 0) continue;
      const names = matchParticipants(m);
      for (let i = 0; i < names.length; i++) {
        if (names[i] !== name || m.kokPaid[i]) continue;
        refs.push({
          kind: "turnamen_kok",
          id: t.id,
          index: i,
          matchId: m.id,
          date: matchPlayedDate(t, m),
          createdAt: t.createdAt,
          amount: perPerson,
        });
      }
    }
  }
  return refs;
}

function byOldest(a: UnpaidRef, b: UnpaidRef): number {
  return (
    String(a.date).localeCompare(String(b.date)) ||
    String(a.createdAt).localeCompare(String(b.createdAt))
  );
}

function slot(r: UnpaidRef): TouchedSlot {
  return { kind: r.kind, id: r.id, index: r.index, ...(r.matchId ? { matchId: r.matchId } : {}) };
}

/**
 * Bayar sebagian: greedy lunasin tagihan yang nominalnya paling pas (best-fit)
 * dulu, bukan yang paling lama. Jadi cicilan 10.000 buat orang yang punya
 * utang kok 3.000 + 2.500 + iuran patungan 10.000 bakal nutup patungan itu
 * pas (diff 0), bukan malah kepecah nyicil kok 5.500 lalu nyisa ganjil di
 * patungan. Tiap langkah pilih ref ber-amount ≤ credit dengan selisih paling
 * kecil; kalau seri, yang paling lama menang (byOldest) biar deterministik.
 */
export function planInstallment(
  games: StoredGame[],
  carry: CarryMap,
  name: string,
  amount: number,
  tournaments: StoredTournament[] = [],
): SettlePlan {
  let credit = Math.max(0, Number(carry[name]) || 0) + amount;
  const refs = unpaidRefs(games, name, tournaments);
  const remaining = [...refs];
  const touched: TouchedSlot[] = [];
  while (true) {
    let bestIdx = -1;
    for (let i = 0; i < remaining.length; i++) {
      const r = remaining[i];
      if (r.amount > credit) continue;
      if (bestIdx === -1) {
        bestIdx = i;
        continue;
      }
      const best = remaining[bestIdx];
      const diff = credit - r.amount;
      const bestDiff = credit - best.amount;
      if (diff < bestDiff || (diff === bestDiff && byOldest(r, best) < 0)) {
        bestIdx = i;
      }
    }
    if (bestIdx === -1) break;
    const r = remaining[bestIdx];
    touched.push(slot(r));
    credit -= r.amount;
    remaining.splice(bestIdx, 1);
  }
  return {
    touched,
    carryAfter: credit > 0 ? credit : null,
    paymentAmount: amount,
    // Cicilan yang menutup semua tagihan tersisa dihitung lunas (sisa kredit jadi titipan).
    clearsDebt: refs.length > 0 && touched.length === refs.length,
  };
}

/** Lunasin semua tagihan orang sekali klik. Tunai = max(0, total − carry titipan). */
export function planSettle(
  games: StoredGame[],
  carry: CarryMap,
  name: string,
  tournaments: StoredTournament[] = [],
): SettlePlan {
  const carryBefore = Math.max(0, Number(carry[name]) || 0);
  const refs = unpaidRefs(games, name, tournaments).sort(byOldest);
  const settled = refs.reduce((s, r) => s + r.amount, 0);
  return {
    touched: refs.map(slot),
    carryAfter: null,
    paymentAmount: Math.max(0, settled - carryBefore),
    clearsDebt: true,
  };
}

File: v3/server/internal/domain/debt.go
// Port dari v2/lib/domain/debt.ts — hutang & cicilan. Planner murni: balik
// daftar slot yang harus di-flip jadi paid + carry baru + nominal payment
// yang dicatat. Repo layer (F2) yang eksekusi tulisannya.
//
// Dua sumber tagihan: main biasa (per pemain per game) dan kok per partai
// turnamen (cuma ditagih ke 4 pemain partai itu). Iuran patungan turnamen
// (fee flat, pool terpisah) SENGAJA tidak ikut di sini — itu ditagih &
// dilunasin sendiri di halaman Turnamen, biar Rekap fokus ke kok aja.
package domain

// BuildDebtSummary — ringkasan hutang per orang: sisa = max(0, owedGross −
// carry).
func BuildDebtSummary(games []EnrichedGame, carry CarryMap, tournaments []EnrichedTournament) []DebtEntry {
	type accum struct {
		name      string
		owedGross int64
		items     []DebtItem
	}
	byName := map[string]*accum{}
	var order []string
	bucket := func(name string) *accum {
		if a, ok := byName[name]; ok {
			return a
		}
		a := &accum{name: name}
		byName[name] = a
		order = append(order, name)
		return a
	}

	matchNumbers := GameMatchNumbers(storedGamesOf(games))

	for _, g := range games {
		for _, p := range g.Players {
			if p.Paid || p.Name == "" {
				continue
			}
			e := bucket(p.Name)
			e.owedGross += g.Cost.PerPerson
			e.items = append(e.items, DebtItem{
				GameID:      g.ID,
				Date:        g.Date,
				Name:        p.Name,
				Amount:      g.Cost.PerPerson,
				KokCount:    g.Cost.KokCount,
				Kind:        DebtKindGame,
				CreatedAt:   g.CreatedAt,
				MatchNumber: matchNumbers[g.ID],
			})
		}
	}

	for _, tt := range tournaments {
		for _, m := range tt.Matches {
			if m.KokTotal <= 0 {
				continue
			}
			perPerson := MatchKokPerPerson(m)
			if perPerson <= 0 {
				continue
			}
			names := MatchParticipants(m)
			for i, name := range names {
				if name == "" || m.KokPaid[i] {
					continue
				}
				e := bucket(name)
				e.owedGross += perPerson
				e.items = append(e.items, DebtItem{
					GameID:    tt.ID + "::" + m.ID,
					Date:      MatchPlayedDate(tt.StoredTournament, m),
					Name:      name,
					Amount:    perPerson,
					KokCount:  len(m.Koks),
					Kind:      DebtKindTurnamenKok,
					Label:     tt.Name + " · " + MatchTitle(tt.Format, tt.Size, m.Round, m.Index),
					CreatedAt: tt.CreatedAt,
				})
			}
		}
	}

	out := make([]DebtEntry, 0, len(order))
	for _, name := range order {
		a := byName[name]
		c := carry[name]
		if c < 0 {
			c = 0
		}
		total := a.owedGross - c
		if total < 0 {
			total = 0
		}
		out = append(out, DebtEntry{Name: a.name, OwedGross: a.owedGross, Items: a.items, Carry: c, Total: total})
	}
	sortDebtEntries(out)
	return out
}

func sortDebtEntries(entries []DebtEntry) {
	// insertion sort — jumlah entri sekecil jumlah anggota klub (§4.2: tidak
	// ada agregat besar di memori); ganti sort.Slice kalau nanti perlu.
	for i := 1; i < len(entries); i++ {
		j := i
		for j > 0 && less(entries[j], entries[j-1]) {
			entries[j-1], entries[j] = entries[j], entries[j-1]
			j--
		}
	}
}

// less: urut total menurun, seri dipecah nama menaik.
func less(a, b DebtEntry) bool {
	if a.Total != b.Total {
		return a.Total > b.Total
	}
	return a.Name < b.Name
}

func storedGamesOf(games []EnrichedGame) []StoredGame {
	out := make([]StoredGame, len(games))
	for i, g := range games {
		out[i] = g.StoredGame
	}
	return out
}

// --- Cicilan & lunas ---

type unpaidRef struct {
	slot      TouchedSlot
	date      string
	createdAt string
	amount    int64
}

func (r unpaidRef) orderKey() string { return r.date + "\x00" + r.createdAt }

func unpaidRefs(games []StoredGame, name string, tournaments []StoredTournament) []unpaidRef {
	var refs []unpaidRef
	for _, g := range games {
		perPerson := GameCostOf(g.Koks, len(g.Players)).PerPerson
		for i, p := range g.Players {
			if p.Name == name && !p.Paid {
				refs = append(refs, unpaidRef{
					slot:      TouchedSlot{Kind: DebtKindGame, ID: g.ID, Index: i},
					date:      g.Date,
					createdAt: g.CreatedAt,
					amount:    perPerson,
				})
			}
		}
	}
	for _, t := range tournaments {
		for _, m := range TournamentMatches(t) {
			if m.KokTotal <= 0 {
				continue
			}
			perPerson := MatchKokPerPerson(m)
			if perPerson <= 0 {
				continue
			}
			names := MatchParticipants(m)
			for i, n := range names {
				if n != name || m.KokPaid[i] {
					continue
				}
				refs = append(refs, unpaidRef{
					slot:      TouchedSlot{Kind: DebtKindTurnamenKok, ID: t.ID, Index: i, MatchID: m.ID},
					date:      MatchPlayedDate(t, m),
					createdAt: t.CreatedAt,
					amount:    perPerson,
				})
			}
		}
	}
	return refs
}

// PlanInstallment — bayar sebagian: greedy lunasin tagihan yang nominalnya
// paling pas (best-fit) dulu, bukan yang paling lama (§9.5 B).
func PlanInstallment(games []StoredGame, carry CarryMap, name string, amount int64, tournaments []StoredTournament) SettlePlan {
	c := carry[name]
	if c < 0 {
		c = 0
	}
	credit := c + amount
	refs := unpaidRefs(games, name, tournaments)

	touched, remaining := bestFitPlan(refs,
		func(r unpaidRef) int64 { return r.amount },
		unpaidRef.orderKey,
		credit,
	)

	slots := make([]TouchedSlot, len(touched))
	for i, r := range touched {
		slots[i] = r.slot
	}

	plan := SettlePlan{
		Touched:       slots,
		PaymentAmount: amount,
		// Cicilan yang menutup semua tagihan tersisa dihitung lunas (sisa
		// kredit jadi titipan).
		ClearsDebt: len(refs) > 0 && len(touched) == len(refs),
	}
	if remaining > 0 {
		plan.CarryAfter = remaining
		plan.HasCarryAfter = true
	}
	return plan
}

// PlanSettle — lunasin semua tagihan orang sekali klik. Tunai = max(0, total − carry titipan).
func PlanSettle(games []StoredGame, carry CarryMap, name string, tournaments []StoredTournament) SettlePlan {
	carryBefore := carry[name]
	if carryBefore < 0 {
		carryBefore = 0
	}
	refs := unpaidRefs(games, name, tournaments)
	sortUnpaidRefsByOldest(refs)

	var settled int64
	slots := make([]TouchedSlot, len(refs))
	for i, r := range refs {
		settled += r.amount
		slots[i] = r.slot
	}

	payment := settled - carryBefore
	if payment < 0 {
		payment = 0
	}
	return SettlePlan{Touched: slots, PaymentAmount: payment, ClearsDebt: true}
}

func sortUnpaidRefsByOldest(refs []unpaidRef) {
	for i := 1; i < len(refs); i++ {
		j := i
		for j > 0 && refs[j-1].orderKey() > refs[j].orderKey() {
			refs[j-1], refs[j] = refs[j], refs[j-1]
			j--
		}
	}
}
